doc/developer/design/20230418_persist_push_based_updates.md

```diff
@@ -0,0 +1,286 @@
+message ProtoPushDiff {
+    string shard_id = 1;
+    uint64 seqno = 2;
+    bytes diff = 3;
+}
+
+message ProtoPushSubscribe {
+  repeated string shards = 1;
+}
+
+message ProtoPushMessage {
+  oneof message {
+    ProtoPushDiff push_diff = 1;
+    ProtoPushSubscribe subscribe = 2;
+  }
+}
+
+service ProtoPersist {
+    rpc Push (stream ProtoPushMessage) returns (stream ProtoPushMessage);
+}
+```
+
+The proposed topology would have each `clusterd` connect to `environmentd`'s RPC service. `clusterd` would subscribe to
+updates to any shards in its `PersistClientCache` (a singleton in both `environmentd` and `clusterd`). When any `clusterd`
+commits a change to a shard, it publishes the diff to `environmentd`, which tracks which connections are subscribed to which
+shards, and broadcasts out the diff to the interested subscribers. From there, each subscriber would [apply the diff](#applying-received-updates).
+
+`environmentd` is chosen to host the RPC service as it necessarily holds a handle to each shard (and therefore is always
+interested in all diffs), and to align with the existing topology of the Controller, rather than introduce the novel
+complexity of point-to-point connections between unrelated `clusterd` processes or a net-new dedicated process.
+
+With this implementation we would anticipate the end-to-end latency between a writer in one `clusterd` pushing an update
+to a reader in another `clusterd` receiving that update to be in the 1-10ms range.
+
+# Rollout
+[rollout]: #rollout
+
+We will roll out this change in phases behind a feature flag. The change is seen purely as an optimization opportunity,
+and should be safe to shut off at any time. We will aim ensure that any errors handled due to Persist PubSub cannot
+crash or halt the process.
+
+We anticipate message volume to be on the order of 1-1.5 per second per shard, with each message in the 100 bytes-1KiB
+range. Given the scale of our usage today, we would expect `environmentd` to be able to broadcast this data and message
+volume comfortably with minimal impact on performance.
+
+That said, we'll want to keep an eye on how much time `environmentd` spends broadcasting diffs, as it will be a new
+workload for the process. We anticipate `clusterd` to be negligibly impacted by this added RPC traffic, as each `clusterd`
+would only need to push/receive updates for the shards it is reading and writing, work is that is very similar to what
+it is doing today.
+
+Our existing Persist metrics dashboard, in addition to the new metrics outlined below, should be sufficient to monitor
+the change and understand its impact.
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+We will introduce several metrics to understand the behavior of the change:
+
+* # of updates pushed
+* # of updates received that apply cleanly & uncleanly
+* How frequently `Listen`s need to fallback to polling
+* End-to-end latency of appending data to observing it
```

Comment:
was thinking of tagging messages with a CaS commit timestamp and plumbing that through. it's imperfect because clock skew etc but could be a decent ballpark

Reply:
we already put the timestamp on diffs! :D

and yeah, that seems fine. i was mostly curious if you'd come up with something clever to account for the clock offsets